### DISCUSSION TITLE
The problem when I install HighFive by cmake

### DISCUSSION THREAD

**Discussion by dirichlet10** [BlueBrain/HighFive#1047] [Questions] [answered]
I follow the steps of "Regular Installation of [HighFive]" on https://bluebrain.github.io/HighFive/ to install HighFive by cmake. I type in the following command:

```
cmake -DHIGHFIVE_EXAMPLES=Off -DHIGHFIVE_USE_BOOST=Off -DHIGHFIVE_UNIT_TESTS=Off -DCMAKE_INSTALL_PREFIX="${HOME}/usr/" -B ./ ../
```

I am in the file "build" under the file "HighFive-src".
But the result shows that:

```
CMake Error at /cluster/home/gengl/usr/lib/cmake/HighFive/HighFiveConfig.cmake:18 (add_library):
  add_library cannot create ALIAS target "HighFive" because target
  "HighFive::HighFive" is itself an ALIAS.
Call Stack (most recent call first):
  CMakeLists.txt:131 (find_package)

CMake Error at /cluster/home/gengl/usr/lib/cmake/HighFive/HighFiveConfig.cmake:19 (add_library):
  add_library cannot create ALIAS target "HighFiveInclude" because target
  "HighFive::Include" is itself an ALIAS.
Call Stack (most recent call first):
  CMakeLists.txt:131 (find_package)

-- Configuring incomplete, errors occurred!
See also "/cluster/home/gengl/software/HighFive-src/build/CMakeFiles/CMakeOutput.log".
```

Could anyone can tell me the reason, please? Thank you in advance！

**Comment by 1uc** [comment]
The commands doesn't look wrong; and fortunately, I can't reproduce the issue using the latest version of HighFive and some reasonably modern version of CMake.

CMake has a tendency to create certain files/directories that then prevent correct runs later on. Check `git status` if there's any new files/directories, outside of `build`, that look like CMake created them.

First we should make sure that it's a clean copy of HighFive. For example run the following in your home directory:
```
mkdir -p ${HOME}/sandbox/hi5-issue-1047
cd ${HOME}/sandbox/hi5-issue-1047
git clone https://github.com/BlueBrain/HighFive.git HighFive-src
cd HighFive-src
```
Now the exact command you posted, should work, i.e.
```
mkdir build && cd build
cmake -DHIGHFIVE_EXAMPLES=Off -DHIGHFIVE_UNIT_TESTS=Off -DCMAKE_INSTALL_PREFIX="${HOME}/usr/" -B ./ ../
```
(I've remove the flag related to `BOOST` but it doesn't matter.)

Hopefully, using a clean directory fixes the issue, if not, we need to understand why the "installed" copy is involved during configure time:
```
CMake Error at /cluster/home/gengl/usr/lib/cmake/HighFive/HighFiveConfig.cmake:19 (add_library):
```
Note that the file it's complaining about is no located in your source or build directory. It's located in the install directory. You could start by checking:
```
which cmake
alias | grep cmake
```

**Comment by dirichlet10** [comment]
Thank you for your reply! And Sorry I forget to mention one thing, that is only when I add the following sentences in the CMakeLists.txt, the problem will occur:
```
# ...
add_executable(foo foo.cpp)
 
find_package(HighFive REQUIRED)
target_link_libraries(foo HighFive)
```

If I don't add them, there is no problem.

**Reply by 1uc** [reply, marked as answer]
Are you editing `HighFive`s CMakeLists.txt? If no, ignore everything that follows.

If yes, I could see how that would result in the error you're describing. The problem is that inside `HighFive/CMakeLists.txt` we create the targets `HighFive::HighFive` and `HighFive::Include`. The `find_package(HighFive REQUIRED)` instructs CMake to go look for `HighFive`, it finds one in the install directory. It then asks the installed version to create the targets, which already exist. Therefore, it fails.

There's two options:
1. You're developing HighFive. In that case just leave out the `find_package(HighFive REQUIRED)`.
2. You're developing an independent application, but are reusing the HighFive's CMake for some reason, maybe convenience. If that's the case, I'd recommend you move your own code into it's own directory (and own Git repository). Then you create a simple `CMakeLists.txt`. You could start out with something like:
https://cliutils.gitlab.io/modern-cmake/chapters/basics/example.html
    (I quite like that resource for modern CMake.)

    You might need to add a minimum C++ version (possibly in a variation that uses targets):
    ```
    set(CMAKE_CXX_STANDARD 14)
    set(CMAKE_CXX_STANDARD_REQUIRED ON)
    set(CMAKE_CXX_EXTENSIONS OFF)
    ```

As part of our CI we check that HighFive can be used by external projects. The `CMakeLists.txt` for applications is here:
https://github.com/BlueBrain/HighFive/blob/9cbd8349b24c6225f001187cd4d2b2c052b8ffe3/tests/cmake_integration/application/CMakeLists.txt

You'll need to delete very liberally, more than 90% of that file isn't needed after making some choices.

**Reply by dirichlet10** [reply]
Thank you very much! I think this is the reason!

**Comment by 1uc** [comment]
I think I'll close, because I suspect it's too specific to be useful to others.
